patches/minecraft/net/minecraft/world/food/FoodData.java.patch

```diff
@@ -0,0 +1,145 @@
+    public FoodData() {
++      this(null);
++   }
++
++   public FoodData(Player player) {
+       this.f_38697_ = 5.0F;
++      this.player = player;
+    }
+ 
++   @Deprecated // Forge: Don't call this unless the player is not consuming food. Use eat(Item, ItemStack) instead.
+    public void m_38707_(int p_38708_, float p_38709_) {
+-      this.f_38696_ = Math.min(p_38708_ + this.f_38696_, 20);
++      if (net.minecraftforge.event.ForgeEventFactory.onFoodDataAddition(this.player, new net.minecraftforge.common.hunger.FoodValues(p_38708_, p_38709_))) return;
++      this.f_38696_ = Math.min(p_38708_ + this.f_38696_, net.minecraftforge.event.ForgeEventFactory.getMaxHunger(this.player));
+       this.f_38697_ = Math.min(this.f_38697_ + (float)p_38708_ * p_38709_ * 2.0F, (float)this.f_38696_);
+    }
+ 
+    public void m_38712_(Item p_38713_, ItemStack p_38714_) {
+-      if (p_38713_.m_41472_()) {
+-         FoodProperties foodproperties = p_38713_.m_41473_();
+-         this.m_38707_(foodproperties.m_38744_(), foodproperties.m_38745_());
++      net.minecraftforge.common.hunger.FoodValues modifiedFoodValues = net.minecraftforge.common.hunger.FoodValues.get(p_38714_, this.player);
++      if (modifiedFoodValues != null) {
++         int prevFoodLevel = this.f_38696_;
++         float prevSaturationLevel = this.f_38697_;
++         this.m_38707_(modifiedFoodValues.getNutrition(), modifiedFoodValues.getSaturationModifier());
++         net.minecraftforge.event.ForgeEventFactory.onFoodEaten(modifiedFoodValues, this.f_38696_ - prevFoodLevel, this.f_38697_ - prevSaturationLevel, p_38714_, this.player);
+       }
+ 
+    }
+ 
+    public void m_38710_(Player p_38711_) {
+-      Difficulty difficulty = p_38711_.f_19853_.m_46791_();
++      if (this.player == null) this.player = p_38711_; // Forge: Safety-check, in-case the class was constructed without the player-specific constructor
+       this.f_38700_ = this.f_38696_;
+-      if (this.f_38698_ > 4.0F) {
+-         this.f_38698_ -= 4.0F;
+-         if (this.f_38697_ > 0.0F) {
+-            this.f_38697_ = Math.max(this.f_38697_ - 1.0F, 0.0F);
+-         } else if (difficulty != Difficulty.PEACEFUL) {
+-            this.f_38696_ = Math.max(this.f_38696_ - 1, 0);
++      net.minecraftforge.eventbus.api.Event.Result allowExhaustionResult = net.minecraftforge.event.ForgeEventFactory.fireAllowExhaustionEvent(p_38711_);
++      float maxExhaustion = net.minecraftforge.event.ForgeEventFactory.getMaxExhaustion(p_38711_);
++      if (allowExhaustionResult == net.minecraftforge.eventbus.api.Event.Result.ALLOW || (allowExhaustionResult == net.minecraftforge.eventbus.api.Event.Result.DEFAULT && this.f_38698_ > maxExhaustion)) {
++         net.minecraftforge.event.hunger.ExhaustionEvent.Exhausted exhaustedEvent = net.minecraftforge.event.ForgeEventFactory.onExhausted(p_38711_, maxExhaustion, this.f_38698_);
++         this.f_38698_ += exhaustedEvent.getDeltaExhaustion();
```

Comment:
Moving this line into the if block would make the behavior the same as if the mod listened to `ExhaustionEvent.AllowExhaustion` and set that to `DENY`. The purpose of the `Exhausted` event is more to be a notification that max exhaustion was reached, while also allowing control over how hunger and saturation values are changed because of the exhaustion occurred. Because of this, unless there's some data `Exhausted` has that `AllowExhaustion` doesn't and should, I think it's better to leave it as is.

Reply:
It is reasonable that cancelling exhaustion have the same impact as returning `DENY` from `AllowExhaustion`, as denying it being allowed, and cancelling it from happening, should technically be the same thing.

Reply:
While that could be reasonable, I'm not sure how I feel about that from a design perspective. Having that basically yields us the below:
```java
if (eventA) {
  if (eventB) {
    doWork();
  }
}
```
To me when I see that, the first thing I think is that there's a chance that one of these events isn't needed, since firing to events one after another to basically ask the same question doesn't seem right. While I'm not against asking that question and maybe we don't actually need to fire both events (I technically didn't write this system so I'd have to do some research as to why both were created in the first place), I think if we do need both events, then I think not moving the line is probably the best course of action. I can't really think of a use case where a modder would say "I do want to have `AllowExhaustion` to go through, but I want to cancel any further action in `Exhausted`". I agree that is does look a little odd having one action from the event go through before checking if it was canceled, so maybe there's a better way to do what is attempting to be done here. I'm not sure. But unless I'm missing something, I don't think the answer is moving the line into the if-check.

If that is what really is wanted in order to be merged, I guess I don't mind doing it. Just wanted to throw in my perspective on this.

Reply:
See - Initially I was going to write that - why can't we just ditch `AllowExhaustion`
However, the case for `AllowExhaustion` is typically going to be that of a mod which wants to trigger exhaustion for some reason or another.  The use case of the other event would likely be a mod not interacting with `AllowExhaustion` at all, but wants to optionally do something when an exhaustion occurs.  
Semantically, it makes sense that cancelling `Exhausted` would cause the same impact as denying `AllowExhaustion`, but these are different time points, and would likely be different mods interacting in some fashion
Say mod A listens to `AllowExhaustion` and sets it to allow every 2 ticks while using some item, and mod B only listens to `Exhausted` to prevent exhaustion while under some potion effect.  You would expect that mod B only needs to cancel `Exhausted` to trigger no changes, but they would have to cancel it and also set getDeltaExhaustion to zero.

Reply:
I kinda follow you for the most part, save for your example:
> mod B only listens to `Exhausted` to prevent exhaustion while under some potion effect

That's basically the exact contract of the `AllowExhaustion` event. Mod B wants to disallow "full exhaustion" processing to happen while under some potion effect, so they would listen to that and `DENY` it. They don't need to listen to `Exhausted` to prevent it, they need to listen to `AllowExhaustion` and then _not_ allow it. To me, that's just misusing the system as it currently exists.

I would agree that if mod B had to cancel and set `deltaExhaustion` to 0 as the only option to get that desired behavior, that would be bad/counter-intuitive. But for this specific example, they don't have to do that. The current contracts of the events is that `AllowExhaustion` is asking "should I go through the 'player is fully exhausted' processing or not?", whereas `Exhausted` is asking "hey, the player is fully exhausted, here's how much exhaustion is going to be removed (or added). How should we handle the hunger/saturation that should be removed (or added)?". The most correct option if a modder wanted to not do any exhaustion processing, which as your suggesting can be achieved by cancelling `Exhausted` and setting `deltaExhaustion` to 0, would actually be done by `DENY`ing `AllowExhaustion` and not even allowing `Exhausted` to be posted (saving a tiny bit of processing with it). The use case of someone not wanting to listen to `AllowExhaustion` but instead cancel `Exhausted` (and setting `deltaExhaustion` to 0, either explicitly or automatically as you are proposing) is the part that still isn't clicking with me.

Perhaps a different example will click better with me, because part of your reasoning I agree with, but I'm not sure I'm fully sold yet.

Reply:
I think that resonates the actual purpose correctly, so it will be fine as-is.

However, I think we should clamp this subtraction here so the value cannot go below zero.